File: implement-shell-tools/cat/cat.py
#!/usr/bin/env python3
import sys

USAGE = "usage: cat [-belnstuv] [file ...]"
KNOWN_FLAGS = set('belnstuv')


def print_lines(f, flag_n, flag_b):
    """Write lines from file-like object f to stdout, with optional numbering."""
    line_num = 0
    for line in f:
        is_blank = line.rstrip('\n') == ''
        if flag_b and not is_blank:
            line_num += 1
            sys.stdout.write(f"{line_num:6}\t{line}")
        elif flag_n and not flag_b:
            line_num += 1
            sys.stdout.write(f"{line_num:6}\t{line}")
        else:
            sys.stdout.write(line)


def parse_args(args):
    flag_n = False
    flag_b = False
    files = []

    for arg in args:
        if arg.startswith('-') and len(arg) > 1:
            for ch in arg[1:]:
                if ch == 'n':
                    flag_n = True
                elif ch == 'b':
                    flag_b = True
                elif ch not in KNOWN_FLAGS:
                    print(f"cat: illegal option -- {ch}", file=sys.stderr)
                    print(USAGE, file=sys.stderr)
                    sys.exit(1)
        else:
            files.append(arg)

    return flag_n, flag_b, files


def main():
    flag_n, flag_b, files = parse_args(sys.argv[1:])
    exit_code = 0

    try:
        if not files:
            print_lines(sys.stdin, flag_n, flag_b)
        else:
            for filepath in files:
                try:
                    with open(filepath) as f:
                        print_lines(f, flag_n, flag_b)
                    sys.stdout.flush()
                except OSError as e:
                    sys.stdout.flush()
                    print(f"cat: {filepath}: {e.strerror}", file=sys.stderr)
                    exit_code = 1
    except KeyboardInterrupt:
        exit_code = 1

    sys.exit(exit_code)


if __name__ == '__main__':
    main()

File: implement-shell-tools/cat/cat.js
#!/usr/bin/env node
// Usage: node cat.js [-n] [-b] file1 [file2 ...]
const fs = require('fs');

const args = process.argv.slice(2);
let flagN = false;
let flagB = false;
const files = [];

for (const arg of args) {
  if (arg === '-n') flagN = true;
  else if (arg === '-b') flagB = true;
  else files.push(arg);
}

function osError(err) {
  if (err.code === 'EACCES') return 'Permission denied';
  if (err.code === 'ENOENT') return 'No such file or directory';
  return err.message;
}

function processContent(content) {
  const lines = content.split('\n');
  if (lines[lines.length - 1] === '') lines.pop();
  let lineNum = 0;
  for (const line of lines) {
    if (flagB) {
      if (line === '') {
        process.stdout.write('\n');
      } else {
        lineNum++;
        process.stdout.write(`${String(lineNum).padStart(6)}\t${line}\n`);
      }
    } else if (flagN) {
      lineNum++;
      process.stdout.write(`${String(lineNum).padStart(6)}\t${line}\n`);
    } else {
      process.stdout.write(line + '\n');
    }
  }
}

if (files.length === 0) {
  let content = '';
  process.stdin.setEncoding('utf8');
  process.stdin.on('data', chunk => { content += chunk; });
  process.stdin.on('end', () => { processContent(content); });
} else {
  let exitCode = 0;
  for (const filepath of files) {
    try {
      const content = fs.readFileSync(filepath, 'utf8');
      processContent(content);
    } catch (err) {
      process.stderr.write(`cat: ${filepath}: ${osError(err)}\n`);
      exitCode = 1;
    }
  }
  process.exit(exitCode);
}
